drivers/gpu/drm/vc4/vc4_firmware_kms.c

```diff
@@ -1097,6 +1099,17 @@ vc4_crtc_mode_valid(struct drm_crtc *crtc, const struct drm_display_mode *mode)
 		break;
 	}
 
+	/* Pi4 can't generate odd horizontal timings on HDMI, so reject modes
+	 * that would set them.
+	 */
+	if (fkms->bcm2711 &&
+	    (vc4_crtc->display_number == 2 || vc4_crtc->display_number == 7) &&
+	    ((mode->hdisplay & 1) ||			    /* active */
+	     ((mode->hsync_start - mode->hdisplay) & 1) ||  /* front porch */
+	     ((mode->hsync_end - mode->hsync_start) & 1) || /* sync pulse */
+	     ((mode->htotal - mode->hsync_end) & 1)))       /* back porch */
+		return MODE_H_ILLEGAL;
```

Comment:
`(A|B|C|D)&1` should generate better code, but it's arguable which is clearer.

Reply:
I've just done exactly that in the firmware patch. 

It's not a path taken that often, so optimising a few ARM instructions out seems a little excessive. I'll let @pelwell make the call.

Reply:
You could compromise and just change the `||`s to `|`s, which might well generate equally good code.

Reply:
That sounds less readable.
I've gone for the ```(A|B|C|D)&1```

Reply:
Opinions, etc.